app/uk/gov/hmrc/exports/services/mapping/Sanitiser.scala

```diff
@@ -0,0 +1,87 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.services.mapping
+
+import com.sun.xml.bind.marshaller.CharacterEscapeHandler
+
+import java.io.Writer
+import java.lang.Character.{isISOControl, isLetterOrDigit, isWhitespace}
+import scala.annotation.switch
+
+class Sanitiser extends CharacterEscapeHandler {
+
+  /* The default JAXB's handler (MinimumEscapeHandler) for escaping non-XML characters does not
+     take into account most ISO control characters apart from \r and \n. On the contrary raises
+     an exception, which we do not want if we can avoid it just removing them - the user in fact
+     could unintentionally paste one of these control characters from an external source.
+
+     Accordingly, any potential control character, apart from \n, is removed from a declaration's
+     value during the conversion process to XML.
+   */
+  // scalastyle:off cyclomatic.complexity
+  override def escape(array: Array[Char], from: Int, length: Int, isAttVal: Boolean, writer: Writer): Unit =
+    if (length == 1) escape(array(from), isAttVal, writer)
+    else {
+      val sb = new StringBuilder(length)
+      for (ix <- from until length) {
+        val chr = array(ix)
+        (chr: @switch) match {
+          case '\t' => sb += ' '
+          case '\n' => if (isAttVal) sb ++= "&#10;" else sb += ' '
+          case '\r' => if ((ix + 1) < length && array(ix + 1) != '\n') sb += ' '
+          case '\"' => if (isAttVal) sb ++= "&quot;" else sb += '\"'
+          case '&'  => sb ++= "&amp;"
+          case '<'  => sb ++= "&lt;"
+          case '>'  => sb ++= "&gt;"
+          case _ =>
+            if ((chr > 31 && chr < 127) || isLetterOrDigit(chr)) sb += chr
+            else if (!(isWhitespace(chr) || isISOControl(chr))) sb ++= s"&#${chr.toInt};"
+        }
+      }
+
+      writer.write(sb.toCharArray)
+    }
+  // scalastyle:on
+
+  private def escape(chr: Char, isAttVal: Boolean, writer: Writer): Unit =
+    (chr: @switch) match {
+      case '\n' => if (isAttVal) writer.write("&#10;") else writer.write('\n')
+      case '\"' => if (isAttVal) writer.write("&quot;") else writer.write('\"')
+      case '&'  => writer.write("&amp;")
+      case '<'  => writer.write("&lt;")
+      case '>'  => writer.write("&gt;")
+      case _ =>
+        if ((chr > 31 && chr < 127) || isLetterOrDigit(chr)) writer.write(chr.toInt)
+        else if (isWhitespace(chr) || isISOControl(chr)) writer.write(' ')
+        else writer.write(s"&#${chr.toInt};")
+    }
+}
```

Comment:
Why are we duplicating the logic here?

Reply:
Check the `if` condition at line #37. This method handles buffers with a single char. This improves the performance because we do not need to create a new `StringBuilder` for every value

Reply:
Yes I see that, just thinking to avoid duplicating the case matching logic, maybe something like this:

Reply:
Honestly... you try to avoid the duplication (and I'm conscious of that but IMO I do not think it is really important here) whereas I'm more concerned about the performance of the conversion process.

Please consider these improvements over your version:

- I do not create a new String for values consisting of a single char (and we have several of those in a declaration)
- `writer.write(Char)` is more performant than `writer.write(String)`

Also note that after last commit a few `case`s have been removed, so lowering the duplication.


Just my 2 cents :slightly_smiling_face: